### PR TITLE
Add networking interface information to support bundle

### PR DESCRIPTION
**Pull request by christ66**
Add an option to obtain information about the networking interfaces that are part of the computer.

@reviewbybees

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+                while (networkInterfaces.hasMoreElements()) {
+                    pw.println("-----------");
+
+                    NetworkInterface ni = networkInterfaces.nextElement();
+                    pw.println(" * Name " + ni.getDisplayName());
+
+                    byte[] hardwareAddress = ni.getHardwareAddress();
+
+                    // Do not have permissions or address does not exist
+                    if (hardwareAddress == null || hardwareAddress.length == 0)
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: As jenkins may only be bound to the loopback I think we should not break but just not dump the hardware address (or dump something special)

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                        continue;
+
+                    pw.println(" ** Hardware Address - " + Util.toHexString(hardwareAddress));
+                    pw.println(" ** Index - " + ni.getIndex());
+                    pw.println(" ** Inet Address - " + ni.getInetAddresses());
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: This is an array so prints an object :(

`** Inet Address - java.net.NetworkInterface$1checkedAddresses@2ce6ab88`

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        return AsyncResultCache.get(node,
+                networkInterfaceCache,
+                new GetNetworkInterfaces(),
+                "network interfaces",
+                "N/A: No connection to node, or no cache.");
+    }
+
+    private static final class GetNetworkInterfaces implements Callable<String, RuntimeException> {
+        public String call() {
+            StringWriter bos = new StringWriter();
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: seems like an overkill when StringBuilder would work.

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                    byte[] hardwareAddress = ni.getHardwareAddress();
+
+                    // Do not have permissions or address does not exist
+                    if (hardwareAddress != null && hardwareAddress.length != 0)
+                        bos.append(" ** Hardware Address - ").append(Util.toHexString(hardwareAddress)).append("\n");
+
+                    bos.append(" ** Index - ").append(ni.getIndex()).append("\n");
+                    Enumeration<InetAddress> inetAddresses = ni.getInetAddresses();
+                    while (inetAddresses.hasMoreElements()) {
+                        NetworkInterface networkInterface =  networkInterfaces.nextElement();
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: I think you mean to refer to inetAddress here.

**Comment by jtnord** [comment]
:-1: also build failure needs investigating.

**Comment by christ66** [comment]
@jtnord Fixed.

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        return "Networking Interface";
+    }
+
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+                new Content("nodes/master/networkInterface.md") {
+                    @Override
+                    public void writeTo(OutputStream os) throws IOException {
+                        os.write(getNetworkInterface(Jenkins.getInstance()).getBytes("utf-8"));
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: "utf-8" is implementation specific.  Use the mandated "UTF-8" instead.

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                    }
+                }
+        );
+
+        for (final Node node : Jenkins.getInstance().getNodes()) {
+            result.add(
+                    new Content("nodes/slave/" + node.getNodeName() + "/networkInterface.md") {
+                        @Override
+                        public void writeTo(OutputStream os) throws IOException {
+                            os.write(getNetworkInterface(node).getBytes("utf-8"));
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: ditto above.

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        return "Networking Interface";
+    }
+
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+                new Content("nodes/master/networkInterface.md") {
+                    @Override
+                    public void writeTo(OutputStream os) throws IOException {
+                        os.write(getNetworkInterface(Jenkins.getInstance()).getBytes("UTF-8"));
```
Comment (jglick) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: BTW there are at least two `UTF_8` constants: Apache Commons and Guava.

**Comment by jglick** [comment]
:+1:

File: src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java (hunk 7 above)
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/support/impl/NetworkInterfaces.java]: I wouldn't use either of those and wait until JDK7 and use `StandardCharsets` which is even more efficient :-)

**Comment by jtnord** [comment]
:+1:

**Pull request merged**